ghcide/src/Development/IDE/Plugin/CodeAction.hs

```diff
@@ -1385,35 +1406,89 @@ newImportToEdit (unNewImport -> imp) ps fileContents
 -- * If the file has neither existing imports nor a module declaration,
 -- the import will be inserted at line zero if there are no pragmas,
 -- * otherwise inserted one line after the last file-header pragma
-newImportInsertRange :: ParsedSource -> T.Text -> Maybe (Range, Int)
-newImportInsertRange (L _ HsModule {..}) fileContents
+newImportInsertRange :: Annotated ParsedSource -> T.Text -> Maybe (Range, Int)
+newImportInsertRange ps fileContents
   |  Just ((l, c), col) <- case hsmodImports of
-      [] -> findPositionNoImports (fmap reLoc hsmodName) (fmap reLoc hsmodExports) fileContents
-      _  -> findPositionFromImportsOrModuleDecl (map reLoc hsmodImports) last True
+      [] -> (\line -> ((line, 0), 0)) <$> findPositionNoImports ps fileContents
+      _  -> findPositionFromImportsOrModuleDecl (map reLoc hsmodImports) last
   , let insertPos = Position (fromIntegral l) (fromIntegral c)
     = Just (Range insertPos insertPos, col)
   | otherwise = Nothing
+  where
+    L _ HsModule {..} = astA ps
 
 -- | Insert the import under the Module declaration exports if they exist, otherwise just under the module declaration.
 -- If no module declaration exists, then no exports will exist either, in that case
 -- insert the import after any file-header pragmas or at position zero if there are no pragmas
-findPositionNoImports :: Maybe (Located ModuleName) -> Maybe (Located [LIE name]) -> T.Text -> Maybe ((Int, Int), Int)
-findPositionNoImports Nothing _ fileContents = findNextPragmaPosition fileContents
-findPositionNoImports _ (Just hsmodExports) _ = findPositionFromImportsOrModuleDecl hsmodExports id False
-findPositionNoImports (Just hsmodName) _ _ = findPositionFromImportsOrModuleDecl hsmodName id False
+findPositionNoImports :: Annotated ParsedSource -> T.Text -> Maybe Int
```

Comment:
this now does something different, but what? The haddock doesn't make sense, it doesn't insert anything

Reply:
The semantics of this function doesn't change. The existing haddock is saying about where we should insert the new import, rather than actually inserting it. It might be a little misleading though, I will try to reword it.

Reply:
Explaining the context in brief here: previously, the code finding position for new import didn't take the location of `where` into account. We don't have something in the AST describing the module header (`module X (...) where`). Instead, we have the module name and module exports separately. `where` doesn't belong to any of them.

To precisely locate "the end of the module header", we need to find the `where` keyword. That's why using exact-print annotations is a must.